Add `NON_GCP_PRIVATE_IP_PORT` as networkEndpointType for NetworkEndpointGroup

AlexanderEllis opened the pull request:
<!-- Put a description of what this PR is for here, along with any references to issues that this resolves or contributes to -->

This change adds `NON_GCP_PRIVATE_IP_PORT` as a `network_endpoint_type` option for the `google_compute_network_endpoint_group` resource.  This is in support for Hybrid Load Balancing.

I added two examples: one with a basic `google_compute_network_endpoint_group` with the new type, and one with a global external managed load balancer with hybrid connectivity (roughly following the configuration at https://cloud.google.com/load-balancing/docs/https/setting-up-ext-https-hybrid). I also added documentation about the valid combination of this new type and backend service load balancing scheme/balancing modes.

Part of https://github.com/hashicorp/terraform-provider-google/issues/10040

For posterity, this change also sets the `instance` attribute on the `google_compute_network_endpoint` resource to be optional, as Hybrid connectivity NEGs use endpoints with just IP and PORT (without an instance). See https://cloud.google.com/load-balancing/docs/negs#hybrid-neg and https://cloud.google.com/load-balancing/docs/https/setting-up-ext-global-https-hybrid#set_up_the_hybrid_connectivity_neg for more details.


<!--
Replace each [ ] with [X] to check it. Switch to the preview view to make it easier to click on links.
These steps will speed up the review process, and we appreciate you spending time on them before sending
your code to be reviewed.
-->
If this PR is for Terraform, I acknowledge that I have:

- [x] Searched through the [issue tracker](https://github.com/hashicorp/terraform-provider-google/issues) for an open issue that this either resolves or contributes to, commented on it to claim it, and written "fixes {url}" or "part of {url}" in this PR description. If there were no relevant open issues, I opened one and commented that I would like to work on it (not necessary for very small changes).
- [x] [Generated Terraform](https://github.com/GoogleCloudPlatform/magic-modules#generating-downstream-tools), and ran `make test` and `make lint` to ensure it passes unit and linter tests.
- [x] Ensured that all new fields I added that can be set by a user appear in at least one [example](https://github.com/GoogleCloudPlatform/magic-modules/tree/master/templates/terraform/examples) (for generated resources) or [third_party test](https://github.com/GoogleCloudPlatform/magic-modules/tree/master/third_party/terraform/tests) (for handwritten resources or update tests).
- [x] [Ran](https://github.com/hashicorp/terraform-provider-google/blob/master/.github/CONTRIBUTING.md#tests) relevant acceptance tests (If the acceptance tests do not yet pass or you are unable to run them, please let your reviewer know).
- [x] Read the [Release Notes Guide](https://github.com/GoogleCloudPlatform/magic-modules/blob/master/.ci/RELEASE_NOTES_GUIDE.md) before writing my release note below.

<!-- AUTOCHANGELOG for Downstream PRs.

Please select one of the following "release-note:" headings:
    - release-note:enhancement
    - release-note:bug
    - release-note:note
    - release-note:new-resource
    - release-note:new-datasource
    - release-note:deprecation
    - release-note:breaking-change
    - release-note:none
    
Unless you choose release-note:none, please add a release note.

See .ci/RELEASE_NOTES_GUIDE.md for writing good release notes.

You can add more release note blocks if you want more than one CHANGELOG
entry for this PR.
-->
**Release Note Template for Downstream PRs (will be copied)**

```release-note:enhancement
compute: add `NON_GCP_PRIVATE_IP_PORT` value for `network_endpoint_type` in the `google_compute_network_endpoint_group` resource
```

```release-note:enhancement
compute: Update `instance` attribute for `google_compute_network_endpoint` to be optional, as Hybrid connectivity NEGs use network endpoints with just IP and Port.
```

modular-magician commented:
Hello!  I am a robot who works on Magic Modules PRs.

I have detected that you are a community contributor, so your PR will be assigned to someone with a commit-bit on this repo for initial review.

Thanks for your contribution!  A human will be with you soon.

@melinath, please review this PR or find an appropriate assignee.

melinath commented:
/gcbrun

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-5684-old..auto-pr-5684) ( 5 files changed, 335 insertions(+), 5 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-5684-old..auto-pr-5684) ( 5 files changed, 335 insertions(+), 5 deletions(-))
TF Validator: [Diff](https://github.com/modular-magician/terraform-validator/compare/auto-pr-5684-old..auto-pr-5684) ( 2 files changed, 3 insertions(+), 3 deletions(-))
TF OiCS: [Diff](https://github.com/modular-magician/docs-examples/compare/auto-pr-5684-old..auto-pr-5684) ( 8 files changed, 321 insertions(+))

modular-magician commented:
I have triggered VCR tests in RECORDING mode for the following tests that failed during VCR: TestAccDatasourceGoogleServiceNetworkingPeeredDnsDomain_basic|TestAccApigeeEnvironmentIamBindingGenerated|TestAccApigeeEnvironmentIamMemberGenerated|TestAccApigeeEnvironmentIamPolicyGenerated|TestAccBigQueryDataset_basic|TestAccBigQueryDataset_access|TestAccCloudbuildWorkerPool_basic|TestAccComputeGlobalForwardingRule_externalCndLbWithBackendBucketExample|TestAccComputeGlobalForwardingRule_globalForwardingRuleHybridExample|TestAccComputeInstance_guestAccelerator|TestAccComputeInstance_guestAcceleratorSkip|TestAccComputeNetworkEndpointGroup_networkEndpointGroupNonGcpExample|TestAccServiceNetworkingPeeredDNSDomain_basic|TestAccSqlDatabaseInstance_withPrivateNetwork_withAllocatedIpRange You can view the result here: https://ci-oss.hashicorp.engineering/viewQueued.html?itemId=255931

melinath commented:
b/202539913, and more context: b/202525233, b/181008964

melinath commented on mmv1/products/compute/terraform.yaml:
```diff
@@ -1498,6 +1513,12 @@ overrides: !ruby/object:Overrides::ResourceOverrides
+        primary_resource_id: "neg"
+        vars:
+          neg_name: "my-lb-neg"
+          network_name: "neg-network"
```
It looks like `TestAccComputeNetworkEndpointGroup_networkEndpointGroupNonGcpExample` is failing with the following error:

```
Error: "name" ("") doesn't match regexp "^(?:[a-z](?:[-a-z0-9]{0,61}[a-z0-9])?)$"
on terraform_plugin_test.tf line 17, in resource "google_compute_subnetwork" "default":
17:   name          = ""
```

I think you may need to add a `subnetwork_name` variable value here.

AlexanderEllis replied:
Whoops great catch! In all of my forwarding rule excitement, I forgot to go back and fix this test. It actually was a little out of date - subnetworks aren't allowed with non-GCP, so these examples shouldn't have a subnetwork (I updated that for the forwarding rule, but not for this one).

Updated the `network_endpoint_group_non_gcp.tf.erb` file to remove the subnetwork, generated the provider, and ran it:

![image](https://user-images.githubusercontent.com/17228751/153921980-4d6b1805-e102-46a5-b780-d6741b7414b0.png)

Thanks again!

melinath commented on mmv1/products/compute/api.yaml:
```diff
@@ -8155,9 +8155,16 @@ objects:
+          endpoint groups (see https://cloud.google.com/load-balancing/docs/hybrid).
+          Note that NON_GCP_PRIVATE_IP_PORT can only be used with Backend Services
+          that 1) have the following load balancing schemes: EXTERNAL, EXTERNAL_MANAGED,
+          INTERNAL_MANAGED, and INTERNAL_SELF_MANAGED and 2) to support the RATE or 
```
```suggestion
          INTERNAL_MANAGED, and INTERNAL_SELF_MANAGED and 2) support the RATE or 
```

AlexanderEllis replied:
Good catch, thanks!

melinath reviewed (changes requested):
This looks good overall - it looks like one of the new tests is failing. I've noted a potential fix below. If you can test it locally prior to pushing, that would be great!

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-5684-old..auto-pr-5684) ( 5 files changed, 317 insertions(+), 5 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-5684-old..auto-pr-5684) ( 5 files changed, 317 insertions(+), 5 deletions(-))
TF Validator: [Diff](https://github.com/modular-magician/terraform-validator/compare/auto-pr-5684-old..auto-pr-5684) ( 2 files changed, 3 insertions(+), 3 deletions(-))
TF OiCS: [Diff](https://github.com/modular-magician/docs-examples/compare/auto-pr-5684-old..auto-pr-5684) ( 8 files changed, 312 insertions(+))

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-5684-old..auto-pr-5684) ( 5 files changed, 317 insertions(+), 5 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-5684-old..auto-pr-5684) ( 5 files changed, 317 insertions(+), 5 deletions(-))
TF Validator: [Diff](https://github.com/modular-magician/terraform-validator/compare/auto-pr-5684-old..auto-pr-5684) ( 2 files changed, 3 insertions(+), 3 deletions(-))
TF OiCS: [Diff](https://github.com/modular-magician/docs-examples/compare/auto-pr-5684-old..auto-pr-5684) ( 8 files changed, 312 insertions(+))

melinath commented:
The CI/CD tests failed but it looks like for unrelated reasons. I've started a run of just the new tests here: https://ci-oss.hashicorp.engineering/buildConfiguration/GoogleCloudBeta_ProviderGoogleCloudBetaMmUpstream/256720

melinath commented:
also gonna do another /gcbrun for good measure

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-5684-old..auto-pr-5684) ( 5 files changed, 317 insertions(+), 5 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-5684-old..auto-pr-5684) ( 5 files changed, 317 insertions(+), 5 deletions(-))
TF Validator: [Diff](https://github.com/modular-magician/terraform-validator/compare/auto-pr-5684-old..auto-pr-5684) ( 2 files changed, 3 insertions(+), 3 deletions(-))
TF OiCS: [Diff](https://github.com/modular-magician/docs-examples/compare/auto-pr-5684-old..auto-pr-5684) ( 8 files changed, 312 insertions(+))

melinath reviewed (approved):
LGTM - I'll merge in the morning assuming the gcbrun looks good.

ct-dh commented:
I was working on my own version of this but hadn't got round to tidying it up to raise a PR. Does this PR support importing the hybrid NEGs as is? I found I had to make more changes than in this PR to get imports to work correctly (this is the diff I had, but as I say I haven't had a chance to tidy it up, please feel free to grab anything if needed): https://github.com/GoogleCloudPlatform/magic-modules/compare/master...ct-dh:feature/hybrid-neg-support?expand=1

melinath reviewed (changes requested):
Per discussion above, sounds like there are some changes to be made.

AlexanderEllis commented:
Updated NetworkEndpoint instance param to be optional and added an example/tests exercising it. Followed the patterns in `resource_compute_global_network_endpoint` examples for the optional params (custom import and handling). Hat tip to @ct-dh again for the suggestion!

<img width="812" alt="image" src="https://user-images.githubusercontent.com/17228751/154154448-487bafac-fb47-4c15-83d3-341f207a10d5.png">

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-5684-old..auto-pr-5684) ( 7 files changed, 364 insertions(+), 25 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-5684-old..auto-pr-5684) ( 7 files changed, 364 insertions(+), 25 deletions(-))
TF Validator: [Diff](https://github.com/modular-magician/terraform-validator/compare/auto-pr-5684-old..auto-pr-5684) ( 2 files changed, 3 insertions(+), 3 deletions(-))
TF OiCS: [Diff](https://github.com/modular-magician/docs-examples/compare/auto-pr-5684-old..auto-pr-5684) ( 8 files changed, 324 insertions(+))

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-5684-old..auto-pr-5684) ( 7 files changed, 364 insertions(+), 25 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-5684-old..auto-pr-5684) ( 7 files changed, 364 insertions(+), 25 deletions(-))
TF Validator: [Diff](https://github.com/modular-magician/terraform-validator/compare/auto-pr-5684-old..auto-pr-5684) ( 2 files changed, 3 insertions(+), 3 deletions(-))
TF OiCS: [Diff](https://github.com/modular-magician/docs-examples/compare/auto-pr-5684-old..auto-pr-5684) ( 8 files changed, 324 insertions(+))

melinath commented:
New test run for just the new tests: https://ci-oss.hashicorp.engineering/buildConfiguration/GoogleCloudBeta_ProviderGoogleCloudBetaMmUpstream/257226

melinath commented:
/gcbrun

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-5684-old..auto-pr-5684) ( 7 files changed, 364 insertions(+), 25 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-5684-old..auto-pr-5684) ( 7 files changed, 364 insertions(+), 25 deletions(-))
TF Validator: [Diff](https://github.com/modular-magician/terraform-validator/compare/auto-pr-5684-old..auto-pr-5684) ( 2 files changed, 3 insertions(+), 3 deletions(-))
TF OiCS: [Diff](https://github.com/modular-magician/docs-examples/compare/auto-pr-5684-old..auto-pr-5684) ( 8 files changed, 324 insertions(+))

melinath commented:
Trying /gcbrun again

modular-magician commented:
I have triggered VCR tests in RECORDING mode for the following tests that failed during VCR: TestAccDatasourceGoogleServiceNetworkingPeeredDnsDomain_basic|TestAccApigeeEnvironmentIamBindingGenerated|TestAccApigeeEnvironmentIamMemberGenerated|TestAccApigeeEnvironmentIamPolicyGenerated|TestAccCloudbuildWorkerPool_basic|TestAccComputeGlobalForwardingRule_globalForwardingRuleHybridExample|TestAccComputeNetworkEndpointGroup_networkEndpointGroupNonGcpExample|TestAccContainerCluster_withAuthenticatorGroupsConfig|TestAccServiceNetworkingPeeredDNSDomain_basic You can view the result here: https://ci-oss.hashicorp.engineering/viewQueued.html?itemId=257868

melinath commented:
@AlexanderEllis quick question - could you add a note explaining why the `instance` field is being changed to optional?

AlexanderEllis commented:
> @AlexanderEllis quick question - could you add a note explaining why the `instance` field is being changed to optional?

Done: added separate release note and additional links in description for posterity. Thanks!

melinath reviewed (approved):
LGTM - remaining test failures are unrelated.
